cluster-autoscaler/cloudprovider/azure/azure_config.go

```diff
@@ -248,6 +254,15 @@ func BuildAzureConfig(configReader io.Reader) (*Config, error) {
 			cfg.EnableDynamicInstanceList = dynamicInstanceListDefault
 		}
 
+		if getVmssSizeRefreshPeriod := os.Getenv("AZURE_GET_VMSS_SIZE_REFRESH_PERIOD"); getVmssSizeRefreshPeriod != "" {
+			cfg.GetVmssSizeRefreshPeriod, err = strconv.Atoi(getVmssSizeRefreshPeriod)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse AZURE_GET_VMSS_SIZE_REFRESH_PERIOD %q: %v", getVmssSizeRefreshPeriod, err)
+			}
+		} else {
+			cfg.GetVmssSizeRefreshPeriod = VmssSizeRefreshPeriodDefault
+		}
```

Comment:
Hmm. Looks like we're doing the same for others as well like - vmssVmsCacheTTL, vmssVmsCacheTTL, etc

Reply:
No, those lack the "else" part, will preserve config file setting in the absence of env override

Reply:
I assume the cloud config file settings will already have been primed into the `cfg` object? If so this should work

```
...
} else if cfg.GetVmssSizeRefreshPeriod != 0 {
         cfg.GetVmssSizeRefreshPeriod = VmssSizeRefreshPeriodDefault
}
```

Is that what you're thinking @tallaxes?

Reply:
From my understanding, line 127 will result in **it being taken from cloud config file already**.
```go
GetVmssSizeRefreshPeriod int `json:"getVmssSizeRefreshPeriod" yaml:"getVmssSizeRefreshPeriod"`
```

Line 155-162 should take care of the parsing:
```go
if configReader != nil {
	body, err := ioutil.ReadAll(configReader)
	if err != nil {
		return nil, fmt.Errorf("failed to read config: %v", err)
	}
	err = json.Unmarshal(body, cfg)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal config body: %v", err)
	}
}
```	

Then, on line 163 and below, **the environment variables will be used only when cloud config file does not exist**.
```go
else {
	cfg.Cloud = os.Getenv("ARM_CLOUD")
	...
}
```

The problem I am seeing is when cloud config file exists, but does not contain `getVmssSizeRefreshPeriod`, and we set `AZURE_GET_VMSS_SIZE_REFRESH_PERIOD`, then this `AZURE_GET_VMSS_SIZE_REFRESH_PERIOD` will never be used and this configuration will goes to default.

My proposal: consider each variable individually rather than entire file at once, when determining whether to use an environment variable. But not in this PR.

Reply:
> The problem I am seeing is when cloud config file exists, but does not contain getVmssSizeRefreshPeriod, and we set AZURE_GET_VMSS_SIZE_REFRESH_PERIOD, then this AZURE_GET_VMSS_SIZE_REFRESH_PERIOD will never be used and this configuration will goes to default.

That's not true, right?

The logic is:
- if env var is set AND not empty string, then convert string to int and set cfg field
- else: use default

```
if getVmssSizeRefreshPeriod := os.Getenv("AZURE_GET_VMSS_SIZE_REFRESH_PERIOD"); getVmssSizeRefreshPeriod != "" {
	cfg.GetVmssSizeRefreshPeriod, err = strconv.Atoi(getVmssSizeRefreshPeriod)
	if err != nil {
		return nil, fmt.Errorf("failed to parse AZURE_GET_VMSS_SIZE_REFRESH_PERIOD %q: %v", getVmssSizeRefreshPeriod, err)
	}
} else {
	cfg.GetVmssSizeRefreshPeriod = VmssSizeRefreshPeriodDefault
}
```

Reply:
To your second point:
> My proposal: consider each variable individually rather than entire file at once when determining whether to use an environment variable. But not in this PR.

Do you mean:
1. read from config file fully and populate ``cfg
2. then for each variable, check to see if it is already set in `cfg`
3. if it's not, check to see if the corresponding env var is set + set in `cfg` if it is
4. finally, if no `cfg` or env var is set, set it to default

Reply:
the assumption here is that people may set values with a mix of cloud config file and env vars?

Reply:
> That's not true, right?
>
> The logic is:
> 
> if env var is set AND not empty string, then convert string to int and set cfg field
> else: use default

It is true (right?, you may help me check again). All these environment variable parsing, from line 163 to 314, is encased in a giant `else` statement. It will reach this part only if `configReader == nil`, which, if we track it down, indicates that the user never provided the path to the config file.
```go
if configReader != nil {
	body, err := ioutil.ReadAll(configReader)
	if err != nil {
		return nil, fmt.Errorf("failed to read config: %v", err)
	}
	err = json.Unmarshal(body, cfg)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal config body: %v", err)
	}
}
Then, on line 163 and below, the environment variables will be used only when cloud config file does not exist.

else {
	cfg.Cloud = os.Getenv("ARM_CLOUD")
	...
	// What we are looking at
}
```

Reply:
> Do you mean:
> 
> 1. read from config file fully and populate `cfg`
> 2. then for each variable, check to see if it is already set in `cfg`
> 3. if it's not, check to see if the corresponding env var is set + set in `cfg` if it is
> 4. finally, if no `cfg` or env var is set, set it to default

That's one way to go with it---with **config file > env > default**. 
Another way is **env > config file > default**. 

I actually prefer the latter more, given env being more delicate(?) than the config file, as well as having a use case I see in AKS.

Reply:
For now, I'm going to leave the logic as is.

However, want to explicitly call out that we are not handling a key case for most of these variables:
* If config is not nil, but doesn't set a field
* the env var for that field will never be picked up
* the default for that field (if present) will never be picked up

@comtalyst is aiming to address this scenario in his [PR to defork cloud-provider-azure](https://github.com/kubernetes/autoscaler/pull/6947/files#diff-e350455eeabc5662e1fc8162e570b2f45f95699270a30ce51443ba61afa49da6R120-L358)

It's a rather large PR so we may determine that this logic is better to be merged separately. Either way, I think it's fine to keep this field using the same structure as the others and address this in a follow-up PR.